### PR TITLE
Fix inf loop after `restore replica` in the replicated merge tree with zero copy.

### PR DESCRIPTION
**Pull request by MikhailBurdukov**
Without these changes, after the scenario from the test, replication will get stuck in infinite loop with:
```
2024.09.05 13:14:51.488260 [ 631 ] {} <Debug> default.s3_test (45f137df-c1c3-4ec5-987c-31fe01f89ebc): Committing part 20200102_0_0_0 to zookeeper
2024.09.05 13:14:51.489333 [ 631 ] {} <Information> default.s3_test (45f137df-c1c3-4ec5-987c-31fe01f89ebc): The part /clickhouse/zero_copy/zero_copy_s3/45f137df-c1c3-4ec5-987c-31fe01f89ebc/2020010
2_0_0_0/data_smj_lqdkmkvooimafnwqprbtfdhhwkbfr/1 on a replica suddenly appeared, will recheck checksums
2024.09.05 13:14:51.489346 [ 631 ] {} <Debug> default.s3_test (45f137df-c1c3-4ec5-987c-31fe01f89ebc): Committing part 20200102_0_0_0 to zookeeper
2024.09.05 13:14:51.490729 [ 631 ] {} <Information> default.s3_test (45f137df-c1c3-4ec5-987c-31fe01f89ebc): The part /clickhouse/zero_copy/zero_copy_s3/45f137df-c1c3-4ec5-987c-31fe01f89ebc/2020010
2_0_0_0/data_smj_lqdkmkvooimafnwqprbtfdhhwkbfr/1 on a replica suddenly appeared, will recheck checksums
2024.09.05 13:14:51.490751 [ 631 ] {} <Debug> default.s3_test (45f137df-c1c3-4ec5-987c-31fe01f89ebc): Committing part 20200102_0_0_0 to zookeeper
2024.09.05 13:14:51.491770 [ 631 ] {} <Information> default.s3_test (45f137df-c1c3-4ec5-987c-31fe01f89ebc): The part /clickhouse/zero_copy/zero_copy_s3/45f137df-c1c3-4ec5-987c-31fe01f89ebc/2020010
```
Also maybe it make sense to set `replace_zero_copy_lock = true` by default?
https://github.com/ClickHouse/ClickHouse/blob/94a62b1df258199addd50629d94f378253dd71da/src/Storages/StorageReplicatedMergeTree.h#L631

Looks like the same behaviour with the inf loop can still happen:
for example:
https://github.com/ClickHouse/ClickHouse/blob/94a62b1df258199addd50629d94f378253dd71da/src/Storages/MergeTree/MutateFromLogEntryTask.cpp#L244
Or
https://github.com/ClickHouse/ClickHouse/blob/94a62b1df258199addd50629d94f378253dd71da/src/Storages/MergeTree/MergeFromLogEntryTask.cpp#L369 

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix inf loop after `restore replica` in the replicated merge tree with zero copy.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 3d68507b61233a4e3ca45d4e911a3ad8f713009b with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69293/3d68507b61233a4e3ca45d4e911a3ad8f713009b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/2b0fe084ae7d86f27f8f63139e52a6246beee95c/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: tests/integration/test_replicated_merge_tree_s3_zero_copy/test.py [hunk 1]
```diff
@@ -145,9 +146,30 @@ def test_insert_select_replicated(cluster, min_rows_for_wide_part, files_per_par
     ) == (3 * FILES_OVERHEAD) + (files_per_part * 3)
 
 
+def remove_leftovers_from_zk(node_data, node_for_query, replica_name):
```
Comment (MikhailBurdukov) [inline tests/integration/test_replicated_merge_tree_s3_zero_copy/test.py:149]: Trying to stabilise the test, it was falling before the changes. Maybe we should move it in separate PR.

**Comment by MikhailBurdukov** [comment]
Also can we back-port this PR, if it's possible? 

And we will made separate PR with cleaning zero_copy locks in `system drop replica`.

**Review by jkartseva — APPROVED** [review]
LGTM, added a label to trigger backports.

**Comment by MikhailBurdukov** [comment]
@jkartseva I can't see any backports:( Probably we should add `pr-must-backport`? Looks like `pr-must-backport-clould` makes backport only in the clickhouse inc cloud.
